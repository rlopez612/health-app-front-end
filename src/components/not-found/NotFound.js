import React from 'react';
import style from './NotFound.module.css';

/**
 * @name NotFound
 * @description 404 page
 * @return component
 */
const NotFound = () => <h1 className={style.NotFound}>404 Page Not Found</h1>;

export default NotFound;
