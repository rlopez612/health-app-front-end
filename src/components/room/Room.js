import React, { useEffect, useState } from 'react';
import { Redirect, useParams, useHistory } from 'react-router-dom';
import Form from '../form/Form';
import Input from '../input/Input';
import Modal from '../modal/Modal';
import { isValidNumber } from '../../utils/validation';
import Textarea from '../textarea/Textarea';

const dummyRooms = [
  {
    id: '1',
    roomType: 'Queen',
    description: 'desc',
    rate: 200,
    active: false
  },
  {
    id: '2',
    roomType: 'King',
    description: 'desc',
    rate: 300,
    active: true
  },
  {
    id: '3',
    roomType: 'Double Queen',
    description: 'desc',
    rate: 300,
    active: true
  }
];

const Room = props => {
  const { user } = props;
  const history = useHistory();
  const params = useParams();

  const [apiError, setApiError] = useState(false);

  const [room, setRoom] = useState({
    id: null,
    roomType: '',
    description: '',
    rate: '',
    active: true
  });

  const [errors, setErrors] = useState({
    roomType: false,
    rate: false
  });

  useEffect(() => {
    if (params.id) {
      const room = dummyRooms.find(room => room.id === params.id);
      setRoom(room);
    }
  }, [params.id]);

  const handleSubmit = event => {
    event.preventDefault();
    const errors = {
      roomType: false,
      rate: false
    }
    let invalidForm = false;

    if (room.roomType.length < 3) {
      errors.roomType = true;
      invalidForm = true;
    }
    if (!isValidNumber(room.rate)) {
      errors.rate = true;
      invalidForm = true;
    }

    if (!invalidForm) {
      console.log('submit');
      history.push('/rooms');

    } else {
      setErrors(errors);
    }
  }

  const handleChange = (event, input) => {
    if (errors[input]) {
      setErrors({ ...errors, [input]: false })
    }
    setRoom({ ...room, [input]: event.target.value });
  }

  if (user.role !== 'manager') {
    return <Redirect to="/reservations" />
  }

  return (
    <>
      {apiError && <Modal message="Oops something went wrong" reset={() => setApiError(false)} />}
      <Form
        title={params.id ? 'Edit Room' : 'Create Room'}
        action={params.id ? 'Update' : 'Create'}
        onSubmit={handleSubmit}
      >
        <Input
          label="Room Type"
          type="text"
          error={errors.roomType}
          message="Must be at least 3 characters"
          value={room.roomType}
          onChange={(event) => handleChange(event, 'roomType')}
        />
        <Textarea
          label="Description"
          type="text"
          value={room.description}
          onChange={(event) => handleChange(event, 'description')}
        />
        <Input
          label="Rate"
          type="number"
          error={errors.rate}
          message="Must be number greater than zero"
          value={room.rate}
          onChange={(event) => handleChange(event, 'rate')}
        />
      </Form>
    </>
  );
}

export default Room;
