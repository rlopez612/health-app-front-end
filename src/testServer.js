import { rest } from 'msw';
import { setupServer } from 'msw/node';

// Establish mock server for all endpoints with 'happy path' mock data
const server = setupServer(
  
  rest.get('http://localhost:8080/room-types', (req, res, ctx) => (
    res(ctx.json([
      {
        id: 1,
        name: 'Executive Suite',
        description: 'Full business suite with office',
        rate: 99.99,
        active: true
      },
      {
        id: 2,
        name: 'King Double',
        description: 'Double king non-smoking',
        rate: 149.99,
        active: true
      }
    ]))
  )),

  rest.get('http://localhost:8080/room-types/:id', (req, res, ctx) => {
    const { id } = req.params;
    return res(
      ctx.json({
        id,
        name: 'Executive Suite',
        description: 'Full business suite with office',
        rate: 99.99,
        active: true
      })
    );
  }),

  rest.post('http://localhost:8080/room-types', (req, res, ctx) => res(ctx.json()))
);

// Establish requests interception layer before all tests.
beforeAll(() => server.listen());
// Reset handlers back to the server defaults above when overriding for sad path tests
afterEach(() => server.resetHandlers());
// Clean up after all tests are done, closing the mock server
afterAll(() => server.close());

export { rest, server };
