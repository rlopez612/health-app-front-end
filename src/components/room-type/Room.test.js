import React from 'react';
import {
  render, screen, waitFor, fireEvent
} from '@testing-library/react';
import {
  Router, Route, MemoryRouter, BrowserRouter
} from 'react-router-dom';
import ReactDOM from 'react-dom';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createMemoryHistory } from 'history';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { rest, server } from '../../testServer';
import Room from './RoomType';
import NotFound from '../not-found/NotFound';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <BrowserRouter>
      <Room user={{ role: 'manager' }} />
    </BrowserRouter>, div
  );
});

test('Room renders correctly', () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <Room user={{ role: 'manager' }} />
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('should redirect to reservations page when user is not a manager', () => {
  const history = createMemoryHistory();
  render(
    <Router history={history}>
      <Room user={{ role: 'employee' }} />
    </Router>
  );

  expect(history.location.pathname).toBe('/reservations');
});

it('should redirect to NotFound page when server responds with 404 status', async () => {
  // Override default server response from testServer.js to test 'sad path'
  server.use(
    rest.get('http://localhost:8080/room-types/:id', (req, res, ctx) => res(ctx.status(404)))
  );

  render(
    <MemoryRouter initialEntries={['room-types/edit/10']}>
      <Route path="room-types/edit/:id">
        <Room user={{ role: 'manager' }} />
      </Route>
      <NotFound />
    </MemoryRouter>
  );

  const notFoundPage = await waitFor(() => screen.getByRole('heading'));
  expect(notFoundPage).toHaveTextContent('404 Page Not Found');
});

it('should display the edit form with values from server when id param included in url', async () => {
  render(
    <MemoryRouter initialEntries={['room-types/edit/1']}>
      <Route path="room-types/edit/:id">
        <Room user={{ role: 'manager' }} />
      </Route>
    </MemoryRouter>
  );

  const formTitle = screen.getByRole('heading');
  const roomDescription = await waitFor(() => screen.findByText('Full business suite with office'));

  expect(formTitle).toHaveTextContent('Edit Room Type');
  expect(roomDescription).toBeVisible();
});

it('should validate form fields', () => {
  const history = createMemoryHistory();
  render(
    <Router history={history}>
      <Room user={{ role: 'manager' }} />
    </Router>
  );

  // submit form with blank fields
  fireEvent.submit(screen.getByRole('button'));

  expect(screen.getByText('Must be at least 3 characters')).toBeVisible();
  expect(screen.getByText('Must be number greater than zero')).toBeVisible();
  expect(history.location.pathname).not.toBe('/room-types');
});

it('should redirect to the room types page on successful post', async () => {
  const history = createMemoryHistory();
  render(
    <Router history={history}>
      <Room user={{ role: 'manager' }} />
    </Router>
  );

  fireEvent.input(screen.getByLabelText(/Room Type/i), {
    target: { value: 'Manager Special' }
  });

  fireEvent.input(screen.getByLabelText(/Description/i), {
    target: { value: 'Awesome room!' }
  });

  fireEvent.input(screen.getByLabelText(/Rate/i), {
    target: { value: 199.99 }
  });

  fireEvent.submit(screen.getByRole('button'));
  await waitFor(() => expect(history.location.pathname).toBe('/room-types'));
});

it('should display error message when there is a server error', async () => {
  // Override default server response from testServer.js to test 'sad path'
  server.use(
    rest.post('http://localhost:8080/room-types', (req, res, ctx) => res(ctx.status(500)))
  );

  render(<Room user={{ role: 'manager' }} />, { wrapper: BrowserRouter });

  // Fill out the form fields with valid data
  fireEvent.input(screen.getByLabelText(/Room Type/i), {
    target: { value: 'Manager Special' }
  });

  fireEvent.input(screen.getByLabelText(/Description/i), {
    target: { value: 'Awesome room!' }
  });

  fireEvent.input(screen.getByLabelText(/Rate/i), {
    target: { value: 199.99 }
  });

  // Submit form with valid data
  fireEvent.submit(screen.getByRole('button'));

  await waitFor(() => expect(screen.getByText('Oops something went wrong')).toBeVisible());
  // Close modal and clear error
  fireEvent.click(screen.getByRole('button', { name: /ok/i }));
  expect(screen.queryByText('Oops something went wrong')).toBeNull();
});
