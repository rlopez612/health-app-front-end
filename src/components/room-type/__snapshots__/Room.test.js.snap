// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Room renders correctly 1`] = `
<form
  className="Form"
  onSubmit={[Function]}
>
  <h1
    className="Title"
  >
    Edit Room
  </h1>
  <div
    className="Container"
  >
    <label
      className="Label"
    >
      Room Type
    </label>
    <input
      className="Input"
      error={false}
      label="Room Type"
      message="Must be at least 3 characters"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="Container"
  >
    <label
      className="Label"
    >
      Description
    </label>
    <textarea
      className="Textarea"
      cols="20"
      label="Description"
      onChange={[Function]}
      rows="2"
      type="text"
      value=""
    />
  </div>
  <div
    className="Container"
  >
    <label
      className="Label"
    >
      Rate
    </label>
    <input
      className="Input"
      error={false}
      label="Rate"
      message="Must be number greater than zero"
      onChange={[Function]}
      type="number"
      value=""
    />
  </div>
  <div
    className="Container"
  >
    <label
      className="Label"
    >
      Active
    </label>
    <input
      checked={true}
      className="Checkbox"
      label="Active"
      onChange={[Function]}
      type="checkbox"
    />
  </div>
  <button
    className="Button Primary"
    color="Primary"
    type="Submit"
  >
    Update
  </button>
</form>
`;
